storage/range.go

```diff
@@ -805,10 +841,19 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
 		if !committed {
 			// We didn't commit the batch, but advance the last applied index nonetheless.
 			if err := setAppliedIndex(r.rm.Engine(), r.Desc().RaftID, index); err != nil {
-				log.Fatalc(ctx, "could not advance applied index: %s", err)
+				rErr = newReplicaCorruptionError("could not advance applied index", err)
 			}
 			atomic.StoreUint64(&r.appliedIndex, index)
 		}
+		if proto.IsWrite(args) {
+			// No matter the result, add result to the response cache if this
+			// is a write method. This must be done as part of the execution of
+			// raft commands so that every replica maintains the same responses
+			// to continue request idempotence, even if leadership changes.
+			if pErr := r.respCache.PutResponse(args.Header().CmdID, reply); pErr != nil {
```

Comment:
nit: you have a similar pattern on L844, which calls its error variable `err`. Maybe be consistent and rename that one to `sErr` or something?